base/compiler/abstractinterpretation.jl

```diff
@@ -225,6 +211,26 @@ function abstract_call_gf_by_type(interp::AbstractInterpreter, @nospecialize(f),
 
     rettype = from_interprocedural!(rettype, sv, arginfo, conditionals)
 
+    # Also considering inferring the compilation signature for this method, so
+    # it is available to the compiler in case it ends up needing it.
+    if infer_compilation_signature(interp) && 1 == seen == napplicable && rettype !== Any && rettype !== Union{} && !is_removable_if_unused(all_effects)
+        match = applicable[1]::MethodMatch
+        method = match.method
+        sig = match.spec_types
+        mi = specialize_method(match; preexisting=true)
+        if mi !== nothing && !const_prop_methodinstance_heuristic(interp, match, mi::MethodInstance, arginfo, sv)
+            csig = get_compileable_sig(method, sig, match.sparams)
+            if csig !== nothing && csig !== sig
```

Comment:
I may want to pack these condition checks into a separate function, e.g.
```julia
    if infer_compilation_signature(interp)
        compilation_sig = maybe_compilation_signature(applicable, seen, rettype, all_effects, arginfo, sv)
        if compilation_sig !== nothing
            (; method, csig, sparams) = compilation_sig
            ...
        end
    end
```
But `maybe_compilation_signature`'s argument list is so long depending on many local variables of `abstract_call_gf_by_type`, so I'm fine with the current code if you prefer it.